code/game/objects/structures/tank_dispenser.dm

```diff
@@ -5,29 +5,57 @@
 	icon_state = "dispenser"
 	density = 1
 	anchored = 1.0
-	var/oxygentanks = 10
-	var/plasmatanks = 10
-	var/list/oxytanks = list()	//sorry for the similar var names
-	var/list/platanks = list()
+	var/starting_oxygen_tanks = 10 // The starting amount of oxygen tanks the dispenser gets when it's spawned
+	var/starting_plasma_tanks = 10 // Starting amount of plasma tanks
+	var/list/stored_oxygen_tanks = list() // List of currently stored oxygen tanks
+	var/list/stored_plasma_tanks = list() // And plasma tanks
 
 /obj/structure/dispenser/oxygen
-	plasmatanks = 0
+	starting_plasma_tanks = 0
 
 /obj/structure/dispenser/plasma
-	oxygentanks = 0
+	starting_oxygen_tanks = 0
 
 /obj/structure/dispenser/New()
 	..()
+	initialize_tanks()
 	update_icon()
 
+/obj/structure/dispenser/Destroy()
+	..()
+	remove_all_tanks()
+
+/obj/structure/dispenser/proc/remove_all_tanks()
+	for(var/P in stored_plasma_tanks)
+		qdel(P)

```

Comment:
this won't qdel properly; they're still in the list.

You should use the helper proc for getting rid of things in lists: